<a id='intro'></a>
## Introduction

### Dataset Description 

> In this project we have two data sets about Gun usage in USA, the first data(Gun data == gun) is about every state in USA and number of permit to use guns, and the numbers and kinds of guns (guns, long guns) and the statues of it like (returned, rented, private, redumption and prepawn) that is used in that state, since Nov. 1998 to Septemper 2017. The other data set, we have diffrent facts - there are 65 differnt census measurement as value of Fact- not so related to the gun owning. however it provides us some facts are related and we will use them to anylysis of the first data for answering the questions too.


### Question(s) for Analysis
> 1.What census variable or fact value is most associated with high gun per capita per state? Ceusus data includes state as variable, . 

> 2.Which states have had the highest growth and the lowest growth in gun registrations from Apr 2010 to Jul 2016?

> 3.What is the overall trend of gun purchases by year or by year and month?

In [73]:
#   plan to use.
import pandas as pd
import numpy as np

<a id='wrangling'></a>
## Data Wrangling

> Upload the data, read it, and see the information of every data set. 

In [122]:
census = pd.read_csv('US_census.csv')
gun = pd.read_csv('gun-data.csv') 

In [123]:
census.head(2)

Fact Fact Note    Alabama  \
0       Population estimates, July 1, 2016,  (V2016)       NaN  4,863,300   
1  Population estimates base, April 1, 2010,  (V2...       NaN  4,780,131   

    Alaska    Arizona   Arkansas  California   Colorado Connecticut Delaware  \
0  741,894  6,931,071  2,988,248  39,250,017  5,540,545   3,576,452  952,065   
1  710,249  6,392,301  2,916,025  37,254,522  5,029,324   3,574,114  897,936   

   ... South Dakota Tennessee       Texas       Utah  Vermont   Virginia  \
0  ...       865454   6651194  27,862,596  3,051,217  624,594  8,411,808   
1  ...       814195   6346298  25,146,100  2,763,888  625,741  8,001,041   

  Washington West Virginia  Wisconsin  Wyoming  
0  7,288,000     1,831,102  5,778,708  585,501  
1  6,724,545     1,853,011  5,687,289  563,767  

[2 rows x 52 columns]

In [76]:
gun.head(2)

month    state   permit  permit_recheck  handgun  long_gun  other  \
0  2017-09  Alabama  16717.0             0.0   5734.0    6320.0  221.0   
1  2017-09   Alaska    209.0             2.0   2320.0    2930.0  219.0   

   multiple  admin  prepawn_handgun  ...  returned_other  rentals_handgun  \
0       317    0.0             15.0  ...             0.0              0.0   
1       160    0.0              5.0  ...             0.0              0.0   

   rentals_long_gun  private_sale_handgun  private_sale_long_gun  \
0               0.0                   9.0                   16.0   
1               0.0                  17.0                   24.0   

   private_sale_other  return_to_seller_handgun  return_to_seller_long_gun  \
0                 3.0                       0.0                        0.0   
1                 1.0                       0.0                        0.0   

   return_to_seller_other  totals  
0                     3.0   32019  
1                     0.0    6303  

[2 rows x 27 columns]

***A.1 ) Check the types of the data***

In [77]:
gun.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12485 entries, 0 to 12484
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      12485 non-null  object 
 1   state                      12485 non-null  object 
 2   permit                     12461 non-null  float64
 3   permit_recheck             1100 non-null   float64
 4   handgun                    12465 non-null  float64
 5   long_gun                   12466 non-null  float64
 6   other                      5500 non-null   float64
 7   multiple                   12485 non-null  int64  
 8   admin                      12462 non-null  float64
 9   prepawn_handgun            10542 non-null  float64
 10  prepawn_long_gun           10540 non-null  float64
 11  prepawn_other              5115 non-null   float64
 12  redemption_handgun         10545 non-null  float64
 13  redemption_long_gun        10544 non-null  flo

**Facts about the data**

* We can see that the columns of the gun data are 27, and 12485 rows.*

* there are a lot of missing values like in the permit recheck, "others", "pre pawn others", "redemption others", "returned hund gun"...." *

* All the floats should be int. and we have to reduce the memory usage* 

# Cleaning Data # 
---------


__Change the type, from float to int.__

In [78]:
gun[gun.select_dtypes('int64').columns] = gun.select_dtypes('int64').astype('int16')

*make sure of the changes*

In [79]:
gun.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12485 entries, 0 to 12484
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      12485 non-null  object 
 1   state                      12485 non-null  object 
 2   permit                     12461 non-null  float64
 3   permit_recheck             1100 non-null   float64
 4   handgun                    12465 non-null  float64
 5   long_gun                   12466 non-null  float64
 6   other                      5500 non-null   float64
 7   multiple                   12485 non-null  int16  
 8   admin                      12462 non-null  float64
 9   prepawn_handgun            10542 non-null  float64
 10  prepawn_long_gun           10540 non-null  float64
 11  prepawn_other              5115 non-null   float64
 12  redemption_handgun         10545 non-null  float64
 13  redemption_long_gun        10544 non-null  flo

In [80]:
gun['permit'] = gun['permit'].astype(pd.Int32Dtype())
gun['handgun'] = gun['handgun'].astype(pd.Int32Dtype())
gun['long_gun'] = gun['long_gun'].astype(pd.Int32Dtype())
gun['other'] = gun['other'].astype(pd.Int32Dtype())
gun['multiple'] = gun['multiple'].astype(pd.Int32Dtype())
gun['admin'] = gun['admin'].astype(pd.Int32Dtype())
gun['prepawn_handgun'] = gun['prepawn_handgun'].astype(pd.Int32Dtype())
gun['prepawn_long_gun'] =gun['prepawn_long_gun'].astype(pd.Int32Dtype())
gun['prepawn_other'] =gun['prepawn_other'].astype(pd.Int32Dtype())
gun['redemption_handgun'] =gun['redemption_handgun'].astype(pd.Int32Dtype())
gun['redemption_long_gun'] = gun ['redemption_long_gun'].astype(pd.Int32Dtype())
gun['redemption_other'] = gun ['redemption_other'].astype(pd.Int32Dtype())
gun['returned_handgun'] = gun ['returned_handgun'].astype(pd.Int32Dtype())
gun['returned_long_gun'] = gun ['returned_long_gun'].astype(pd.Int32Dtype())
gun['returned_other'] = gun ['returned_other'].astype(pd.Int32Dtype())
gun['rentals_handgun'] = gun ['rentals_handgun'].astype(pd.Int32Dtype())
gun['rentals_long_gun'] = gun ['rentals_long_gun'].astype(pd.Int32Dtype())
gun['returned_other'] = gun ['returned_other'].astype(pd.Int32Dtype())
gun['rentals_handgun'] = gun ['rentals_handgun'].astype(pd.Int32Dtype())
gun['rentals_long_gun'] = gun ['rentals_long_gun'].astype(pd.Int32Dtype())
gun['private_sale_handgun'] = gun ['private_sale_handgun'].astype(pd.Int32Dtype())
gun['private_sale_long_gun'] = gun ['private_sale_long_gun'].astype(pd.Int32Dtype())
gun['private_sale_other'] = gun ['private_sale_other'].astype(pd.Int32Dtype())
gun['return_to_seller_handgun'] = gun ['return_to_seller_handgun'].astype(pd.Int32Dtype())
gun['return_to_seller_long_gun'] = gun ['return_to_seller_long_gun'].astype(pd.Int32Dtype())
gun['return_to_seller_other'] = gun ['return_to_seller_other'].astype(pd.Int32Dtype())

In [81]:
gun.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12485 entries, 0 to 12484
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   month                      12485 non-null  object 
 1   state                      12485 non-null  object 
 2   permit                     12461 non-null  Int32  
 3   permit_recheck             1100 non-null   float64
 4   handgun                    12465 non-null  Int32  
 5   long_gun                   12466 non-null  Int32  
 6   other                      5500 non-null   Int32  
 7   multiple                   12485 non-null  Int32  
 8   admin                      12462 non-null  Int32  
 9   prepawn_handgun            10542 non-null  Int32  
 10  prepawn_long_gun           10540 non-null  Int32  
 11  prepawn_other              5115 non-null   Int32  
 12  redemption_handgun         10545 non-null  Int32  
 13  redemption_long_gun        10544 non-null  Int



***B.1 ) Check if there is duplicates***

In [82]:
gun.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
12480    False
12481    False
12482    False
12483    False
12484    False
Length: 12485, dtype: bool

*Or, we can use this line to check effectivly*

In [83]:
sum(gun.duplicated())

0

**Facts, There's no duplicates**

*Now, lets delete the nan in the data. And change memory usage and data type*

In [84]:
gun.isnull().sum()


month                            0
state                            0
permit                          24
permit_recheck               11385
handgun                         20
long_gun                        19
other                         6985
multiple                         0
admin                           23
prepawn_handgun               1943
prepawn_long_gun              1945
prepawn_other                 7370
redemption_handgun            1940
redemption_long_gun           1941
redemption_other              7370
returned_handgun             10285
returned_long_gun            10340
returned_other               10670
rentals_handgun              11495
rentals_long_gun             11660
private_sale_handgun          9735
private_sale_long_gun         9735
private_sale_other            9735
return_to_seller_handgun     10010
return_to_seller_long_gun     9735
return_to_seller_other       10230
totals                           0
dtype: int64

gun= gun.fillna(gun....(), axis=0, inplace=True) -- I have to delete the NA ?? 

In [85]:
gun.head(4)

month     state  permit  permit_recheck  handgun  long_gun  other  \
0  2017-09   Alabama   16717             0.0     5734      6320    221   
1  2017-09    Alaska     209             2.0     2320      2930    219   
2  2017-09   Arizona    5069           382.0    11063      7946    920   
3  2017-09  Arkansas    2935           632.0     4347      6063    165   

   multiple  admin  prepawn_handgun  ...  returned_other  rentals_handgun  \
0       317      0               15  ...               0                0   
1       160      0                5  ...               0                0   
2       631      0               13  ...               0                0   
3       366     51               12  ...               0                0   

   rentals_long_gun  private_sale_handgun  private_sale_long_gun  \
0                 0                     9                     16   
1                 0                    17                     24   
2                 0                    38                     12   
3                 0                    13                     23   

   private_sale_other  return_to_seller_handgun  return_to_seller_long_gun  \
0                   3                         0                          0   
1                   1                         0                          0   
2                   2                         0                          0   
3                   0                         0                          2   

   return_to_seller_other  totals  
0                       3   32019  
1                       0    6303  
2                       0   28394  
3                       1   17747  

[4 rows x 27 columns]

**Extract years and monthes to deal with the data accuretlly** 

In [86]:
gun['year']=gun['month'].apply(lambda x: x.split("-")[0]).astype(int)


In [87]:
gun['year'].unique()


array([2017, 2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007,
       2006, 2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998])

In [88]:
gun['months'] = gun['month'].apply(lambda x: x.split("-")[1]).astype(int)


In [89]:
gun['months'].unique()

array([ 9,  8,  7,  6,  5,  4,  3,  2,  1, 12, 11, 10])

In [90]:
gun.head(1)

month    state  permit  permit_recheck  handgun  long_gun  other  \
0  2017-09  Alabama   16717             0.0     5734      6320    221   

   multiple  admin  prepawn_handgun  ...  rentals_long_gun  \
0       317      0               15  ...                 0   

   private_sale_handgun  private_sale_long_gun  private_sale_other  \
0                     9                     16                   3   

   return_to_seller_handgun  return_to_seller_long_gun  \
0                         0                          0   

   return_to_seller_other  totals  year  months  
0                       3   32019  2017       9  

[1 rows x 29 columns]

In [91]:
gun.shape

(12485, 29)

**A.2 ) Now, lets check the second data set. the info and duplicates.**

In [92]:
census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 52 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Fact            80 non-null     object
 1   Fact Note       28 non-null     object
 2   Alabama         65 non-null     object
 3   Alaska          65 non-null     object
 4   Arizona         65 non-null     object
 5   Arkansas        65 non-null     object
 6   California      65 non-null     object
 7   Colorado        65 non-null     object
 8   Connecticut     65 non-null     object
 9   Delaware        65 non-null     object
 10  Florida         65 non-null     object
 11  Georgia         65 non-null     object
 12  Hawaii          65 non-null     object
 13  Idaho           65 non-null     object
 14  Illinois        65 non-null     object
 15  Indiana         65 non-null     object
 16  Iowa            65 non-null     object
 17  Kansas          65 non-null     object
 18  Kentucky    

In [93]:
 census.head(2)

Fact Fact Note    Alabama  \
0       Population estimates, July 1, 2016,  (V2016)       NaN  4,863,300   
1  Population estimates base, April 1, 2010,  (V2...       NaN  4,780,131   

    Alaska    Arizona   Arkansas  California   Colorado Connecticut Delaware  \
0  741,894  6,931,071  2,988,248  39,250,017  5,540,545   3,576,452  952,065   
1  710,249  6,392,301  2,916,025  37,254,522  5,029,324   3,574,114  897,936   

   ... South Dakota Tennessee       Texas       Utah  Vermont   Virginia  \
0  ...       865454   6651194  27,862,596  3,051,217  624,594  8,411,808   
1  ...       814195   6346298  25,146,100  2,763,888  625,741  8,001,041   

  Washington West Virginia  Wisconsin  Wyoming  
0  7,288,000     1,831,102  5,778,708  585,501  
1  6,724,545     1,853,011  5,687,289  563,767  

[2 rows x 52 columns]

**Facts:**
* All of the values are abjects.
* There are missing values.
* There are 85 raws and 52 columns

In [94]:
sum(census.duplicated())

3

*There are 3 duplicates, we have to drop them*
*And, We will drop the fact notes since it does not have valuable data*

In [95]:
census.drop_duplicates(inplace= True)

In [96]:
sum(census.duplicated())

0

In [97]:
census= census.drop('Fact Note', axis=1)

In [98]:
census.head(4)

Fact    Alabama   Alaska  \
0       Population estimates, July 1, 2016,  (V2016)  4,863,300  741,894   
1  Population estimates base, April 1, 2010,  (V2...  4,780,131  710,249   
2  Population, percent change - April 1, 2010 (es...      1.70%    4.50%   
3                  Population, Census, April 1, 2010  4,779,736  710,231   

     Arizona   Arkansas  California   Colorado Connecticut Delaware  \
0  6,931,071  2,988,248  39,250,017  5,540,545   3,576,452  952,065   
1  6,392,301  2,916,025  37,254,522  5,029,324   3,574,114  897,936   
2      8.40%      2.50%       5.40%     10.20%       0.10%    6.00%   
3  6,392,017  2,915,918  37,253,956  5,029,196   3,574,097  897,934   

      Florida  ... South Dakota Tennessee       Texas       Utah  Vermont  \
0  20,612,439  ...       865454   6651194  27,862,596  3,051,217  624,594   
1  18,804,592  ...       814195   6346298  25,146,100  2,763,888  625,741   
2       9.60%  ...        0.063     0.048      10.80%     10.40%   -0.20%   
3  18,801,310  ...       814180   6346105  25,145,561  2,763,885  625,741   

    Virginia Washington West Virginia  Wisconsin  Wyoming  
0  8,411,808  7,288,000     1,831,102  5,778,708  585,501  
1  8,001,041  6,724,545     1,853,011  5,687,289  563,767  
2      5.10%      8.40%        -1.20%      1.60%    3.90%  
3  8,001,024  6,724,540     1,852,994  5,686,986  563,626  

[4 rows x 51 columns]

*We have to check for the nulls*

In [99]:
census.isnull().sum(). sum()

852

*I can't decide if I have to drope all the nulls. maybe not.*

In [100]:
 census.describe()

Fact    Alabama Alaska  \
count                                             80         65     65   
unique                                            80         65     64   
top     Population estimates, July 1, 2016,  (V2016)  4,863,300  7.30%   
freq                                               1          1      2   

       Arizona Arkansas California Colorado Connecticut Delaware Florida  ...  \
count       65       65         65       65          65       65      65  ...   
unique      64       64         63       64          63       64      64  ...   
top     50.30%   50.90%      6.80%    3.30%       0.10%   51.60%  51.10%  ...   
freq         2        2          2        2           2        2       2  ...   

       South Dakota Tennessee   Texas   Utah Vermont   Virginia Washington  \
count            65        65      65     65      65         65         65   
unique           65        64      64     64      63         65         65   
top          865454     0.048  50.40%  2.50%   1.90%  8,411,808  7,288,000   
freq              1         2       2      2       2          1          1   

       West Virginia  Wisconsin Wyoming  
count             65         65      65  
unique            64         65      64  
top            1.50%  5,778,708   7.10%  
freq               2          1       2  

[4 rows x 51 columns]

In [101]:
states=['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',  
        'Hawaii', 'Idaho', 'Illinois','Indiana', 'Iowa', 'Kansas', 
        'Kentucky', 'Louisiana', 'Maine','Maryland', 'Massachusetts', 
        'Michigan','Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska',
       'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma', 'Oregon','Pennsylvania', 
        'Rhode Island', 'South Carolina','South Dakota', 'Tennessee', 'Texas', 'Utah', 
        'Vermont','Virginia', 'Washington', 'West Virginia','Wisconsin', 'Wyoming']

In [102]:
for state in states:
    census[state].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
    # remove all the non digit character
    census[state]=pd.to_numeric(census[state], downcast='float', errors='ignore')

*We want to manipulate states So we gave them a variable*

## Exploratory Data Analysis


> 1.What census variable or fact value is most associated with high gun per capita per state? Ceusus data includes state as variable

*In order to calculate gun per capita,  the gun totals and population for each state needed to be fetched at first, and the gun and census data needed to be combined. Also, I notice the state in census data is divided as 50 columns, however, in gun data, state is
only one columns which has 46 different state value.*
**And Because of that we will change the raws an columns so we can deal with both data**

In [103]:
census.set_index('Fact',inplace=True)
census_T = census.T.reset_index()

In [104]:
census_T.head(2)

Fact    index  Population estimates, July 1, 2016,  (V2016)  \
0     Alabama                                     4863300.0   
1      Alaska                                      741894.0   

Fact  Population estimates base, April 1, 2010,  (V2016)  \
0                                             4780131.0    
1                                              710249.0    

Fact  Population, percent change - April 1, 2010 (estimates base) to July 1, 2016,  (V2016)  \
0                                                 170.0                                       
1                                                 450.0                                       

Fact  Population, Census, April 1, 2010  \
0                             4779736.0   
1                              710231.0   

Fact  Persons under 5 years, percent, July 1, 2016,  (V2016)  \
0                                                 600.0        
1                                                 730.0        

Fact  Persons under 5 years, percent, April 1, 2010  \
0                                             640.0   
1                                             760.0   

Fact  Persons under 18 years, percent, July 1, 2016,  (V2016)  \
0                                                2260.0         
1                                                2520.0         

Fact  Persons under 18 years, percent, April 1, 2010  \
0                                             2370.0   
1                                             2640.0   

Fact  Persons 65 years and over, percent,  July 1, 2016,  (V2016)  ...  (c)  \
0                                                1610.0            ...  NaN   
1                                                1040.0            ...  NaN   

Fact  Value Flags   -   D   F  FN  NaN   S   X   Z  
0             NaN NaN NaN NaN NaN  NaN NaN NaN NaN  
1             NaN NaN NaN NaN NaN  NaN NaN NaN NaN  

[2 rows x 83 columns]

*Rename the column name from index to state to match the column name in gun data*

In [105]:
census_T.rename(columns={'index':'state'},inplace = True)

*The most of the data is between 2010 to 2016, so we can merge between both data we have to use the inner data*

In [106]:
gun10=gun[gun['year'] == 2016]
gun16=gun[gun['year'] == 2010] 

In [107]:
gun10.head(2)

month    state  permit  permit_recheck  handgun  long_gun  other  \
495  2016-12  Alabama   26996             0.0    13733     17519    376   
496  2016-12   Alaska     254             0.0     4087      3698    243   

     multiple  admin  prepawn_handgun  ...  rentals_long_gun  \
495       533      0               10  ...                 0   
496       191      0                3  ...                 0   

     private_sale_handgun  private_sale_long_gun  private_sale_other  \
495                     2                      2                   1   
496                     0                      0                   0   

     return_to_seller_handgun  return_to_seller_long_gun  \
495                         0                          0   
496                         0                          0   

     return_to_seller_other  totals  year  months  
495                       0   -3522  2016      12  
496                       0    8786  2016      12  

[2 rows x 29 columns]

*Group by gun data by state for 2010 and 2016*


In [108]:
guntotal10= gun10.groupby(['state'])['totals'].sum().reset_index()
guntotal16= gun16.groupby(['state'])['totals'].sum().reset_index()

In [109]:
guntotal16.head()

state  totals
0     Alabama  243071
1      Alaska   65909
2     Arizona  206050
3    Arkansas  191448
4  California   29967

In [110]:
guntotal10.head()

state  totals
0     Alabama -169485
1      Alaska   87647
2     Arizona -108009
3    Arkansas  266014
4  California  -47665

In [111]:
#Rename the dataset to represent different time point 
guntotal10.rename(columns={'totals':'2010_totals'},inplace = True)
guntotal16.rename(columns={'totals':'2016_totals'},inplace = True)

In [112]:
#Merge 2010 and 2016 gun data summary 
guntotal = guntotal16.merge(guntotal10, on='state', how='inner')

*See the result*

In [42]:
guntotal.head(4)

state  2016_totals  2010_totals
0   Alabama       243071      -169485
1    Alaska        65909        87647
2   Arizona       206050      -108009
3  Arkansas       191448       266014

In [113]:
#Merge gun and census data with inner join, by state column 
result = guntotal.merge(census_T, on='state', how='inner')

In [114]:
result.head(4)

state  2016_totals  2010_totals  \
0   Alabama       243071      -169485   
1    Alaska        65909        87647   
2   Arizona       206050      -108009   
3  Arkansas       191448       266014   

   Population estimates, July 1, 2016,  (V2016)  \
0                                     4863300.0   
1                                      741894.0   
2                                     6931071.0   
3                                     2988248.0   

   Population estimates base, April 1, 2010,  (V2016)  \
0                                          4780131.0    
1                                           710249.0    
2                                          6392301.0    
3                                          2916025.0    

   Population, percent change - April 1, 2010 (estimates base) to July 1, 2016,  (V2016)  \
0                                              170.0                                       
1                                              450.0                                       
2                                              840.0                                       
3                                              250.0                                       

   Population, Census, April 1, 2010  \
0                          4779736.0   
1                           710231.0   
2                          6392017.0   
3                          2915918.0   

   Persons under 5 years, percent, July 1, 2016,  (V2016)  \
0                                              600.0        
1                                              730.0        
2                                              630.0        
3                                              640.0        

   Persons under 5 years, percent, April 1, 2010  \
0                                          640.0   
1                                          760.0   
2                                          710.0   
3                                          680.0   

   Persons under 18 years, percent, July 1, 2016,  (V2016)  ...  (c)  \
0                                             2260.0        ...  NaN   
1                                             2520.0        ...  NaN   
2                                             2350.0        ...  NaN   
3                                             2360.0        ...  NaN   

   Value Flags   -   D   F  FN  NaN   S   X   Z  
0          NaN NaN NaN NaN NaN  NaN NaN NaN NaN  
1          NaN NaN NaN NaN NaN  NaN NaN NaN NaN  
2          NaN NaN NaN NaN NaN  NaN NaN NaN NaN  
3          NaN NaN NaN NaN NaN  NaN NaN NaN NaN  

[4 rows x 85 columns]

In [115]:
result['Population estimates, July 1, 2016,  (V2016)'] = result['Population estimates, July 1, 2016,  (V2016)'].replace(',', '').astype(float)

result['Population estimates base, April 1, 2010, (V2016)'] = result['Population estimates base, April 1, 2010,  (V2016)'].replace(',', '').astype(float)

In [116]:
result['Gun_Per_Capital_2016'] = result['2016_totals'] / result['Population estimates, July 1, 2016,  (V2016)']

In [117]:
result['Gun_Per_Capital_2010'] = result['2010_totals'] / result['Population estimates base, April 1, 2010,  (V2016)']

In [118]:
result.head(9)


state  2016_totals  2010_totals  \
0      Alabama       243071      -169485   
1       Alaska        65909        87647   
2      Arizona       206050      -108009   
3     Arkansas       191448       266014   
4   California        29967       -47665   
5     Colorado       235284      -245918   
6  Connecticut       179595       186620   
7     Delaware        19240        59430   
8      Florida      -227085       -71988   

   Population estimates, July 1, 2016,  (V2016)  \
0                                     4863300.0   
1                                      741894.0   
2                                     6931071.0   
3                                     2988248.0   
4                                    39250017.0   
5                                     5540545.0   
6                                     3576452.0   
7                                      952065.0   
8                                    20612439.0   

   Population estimates base, April 1, 2010,  (V2016)  \
0                                          4780131.0    
1                                           710249.0    
2                                          6392301.0    
3                                          2916025.0    
4                                         37254522.0    
5                                          5029324.0    
6                                          3574114.0    
7                                           897936.0    
8                                         18804592.0    

   Population, percent change - April 1, 2010 (estimates base) to July 1, 2016,  (V2016)  \
0                                              170.0                                       
1                                              450.0                                       
2                                              840.0                                       
3                                              250.0                                       
4                                              540.0                                       
5                                             1020.0                                       
6                                               10.0                                       
7                                              600.0                                       
8                                              960.0                                       

   Population, Census, April 1, 2010  \
0                          4779736.0   
1                           710231.0   
2                          6392017.0   
3                          2915918.0   
4                         37253956.0   
5                          5029196.0   
6                          3574097.0   
7                           897934.0   
8                         18801310.0   

   Persons under 5 years, percent, July 1, 2016,  (V2016)  \
0                                              600.0        
1                                              730.0        
2                                              630.0        
3                                              640.0        
4                                              630.0        
5                                              610.0        
6                                              520.0        
7                                              580.0        
8                                              550.0        

   Persons under 5 years, percent, April 1, 2010  \
0                                          640.0   
1                                          760.0   
2                                          710.0   
3                                          680.0   
4                                          680.0   
5                                          680.0   
6                                          570.0   
7                                          620.0   
8                                          570.0   

   Persons under 18 years, perce

the highest state per capital on 2010

In [119]:
result.nlargest(5,'Gun_Per_Capital_2010')


state  2016_totals  2010_totals  \
25        Montana       101095       136337   
40   South Dakota        63151       111921   
47  West Virginia       159550       242350   
28  New Hampshire        79940       165164   
1          Alaska        65909        87647   

    Population estimates, July 1, 2016,  (V2016)  \
25                                     1042520.0   
40                                      865454.0   
47                                     1831102.0   
28                                     1334795.0   
1                                       741894.0   

    Population estimates base, April 1, 2010,  (V2016)  \
25                                           989414.0    
40                                           814195.0    
47                                          1853011.0    
28                                          1316461.0    
1                                            710249.0    

    Population, percent change - April 1, 2010 (estimates base) to July 1, 2016,  (V2016)  \
25                                              540.0                                       
40                                               63.0                                       
47                                              120.0                                       
28                                              140.0                                       
1                                               450.0                                       

    Population, Census, April 1, 2010  \
25                           989415.0   
40                           814180.0   
47                          1852994.0   
28                          1316470.0   
1                            710231.0   

    Persons under 5 years, percent, July 1, 2016,  (V2016)  \
25                                              600.0        
40                                               71.0        
47                                              550.0        
28                                              480.0        
1                                               730.0        

    Persons under 5 years, percent, April 1, 2010  \
25                                          630.0   
40                                           73.0   
47                                          560.0   
28                                          530.0   
1                                           760.0   

    Persons under 18 years, percent, July 1, 2016,  (V2016)  ...   D   F  FN  \
25                                             2180.0        ... NaN NaN NaN   
40                                              246.0        ... NaN NaN NaN   
47                                             2050.0        ... NaN NaN NaN   
28                                             1950.0        ... NaN NaN NaN   
1                                              2520.0        ... NaN NaN NaN   

    NaN   S   X   Z  Population estimates base, April 1, 2010, (V2016)  \
25  NaN NaN NaN NaN                                           989414.0   
40  NaN NaN NaN NaN                                           814195.0   
47  NaN NaN NaN NaN                                          1853011.0   
28  NaN NaN NaN NaN                                          1316461.0   
1   NaN NaN NaN NaN                                           710249.0   

    Gun_Per_Capital_2016  Gun_Per_Capital_2010  
25              0.096972              0.137796  
40              0.072969              0.137462  
47              0.087133              0.130787  
28              0.059889              0.125461  
1               0.088839              0.123403  

[5 rows x 88 columns]

the highest state per capital on 2016

In [120]:
result.nlargest(5,'Gun_Per_Capital_2016')

state  2016_totals  2010_totals  \
25        Montana       101095       136337   
1          Alaska        65909        87647   
47  West Virginia       159550       242350   
49        Wyoming        47709        63594   
40   South Dakota        63151       111921   

    Population estimates, July 1, 2016,  (V2016)  \
25                                     1042520.0   
1                                       741894.0   
47                                     1831102.0   
49                                      585501.0   
40                                      865454.0   

    Population estimates base, April 1, 2010,  (V2016)  \
25                                           989414.0    
1                                            710249.0    
47                                          1853011.0    
49                                           563767.0    
40                                           814195.0    

    Population, percent change - April 1, 2010 (estimates base) to July 1, 2016,  (V2016)  \
25                                              540.0                                       
1                                               450.0                                       
47                                              120.0                                       
49                                              390.0                                       
40                                               63.0                                       

    Population, Census, April 1, 2010  \
25                           989415.0   
1                            710231.0   
47                          1852994.0   
49                           563626.0   
40                           814180.0   

    Persons under 5 years, percent, July 1, 2016,  (V2016)  \
25                                              600.0        
1                                               730.0        
47                                              550.0        
49                                              650.0        
40                                               71.0        

    Persons under 5 years, percent, April 1, 2010  \
25                                          630.0   
1                                           760.0   
47                                          560.0   
49                                          710.0   
40                                           73.0   

    Persons under 18 years, percent, July 1, 2016,  (V2016)  ...   D   F  FN  \
25                                             2180.0        ... NaN NaN NaN   
1                                              2520.0        ... NaN NaN NaN   
47                                             2050.0        ... NaN NaN NaN   
49                                             2370.0        ... NaN NaN NaN   
40                                              246.0        ... NaN NaN NaN   

    NaN   S   X   Z  Population estimates base, April 1, 2010, (V2016)  \
25  NaN NaN NaN NaN                                           989414.0   
1   NaN NaN NaN NaN                                           710249.0   
47  NaN NaN NaN NaN                                          1853011.0   
49  NaN NaN NaN NaN                                           563767.0   
40  NaN NaN NaN NaN                                           814195.0   

    Gun_Per_Capital_2016  Gun_Per_Capital_2010  
25              0.096972              0.137796  
1               0.088839              0.123403  
47              0.087133              0.130787  
49              0.081484              0.112802  
40              0.072969              0.137462  

[5 rows x 88 columns]

In [121]:
fact = result.drop(['Gun_Per_Capital_2010','state','FIPS Code','2016_totals','2010_totals'],axis=1)


# Plotting

In [72]:
result.plot()

<class 'ImportError'>: matplotlib is required for plotting when the default backend "matplotlib" is selected.

In [ ]:
plots_created = False  # Flag to check if any plots were created

for col in fact.columns:
    # Ensure the column name is a string before checking for 'firms'
    col_str = str(col)
    if 'firms' in col_str:
        plt.scatter(fact[col], fact['Gun_Per_Capital_2016'], label=col_str)
        plots_created = True

if plots_created:
    plt.xlabel('Firms')
    plt.ylabel('Gun Per Capital 2016')
    plt.legend()
    plt.show()
else:
    print("No columns containing 'firms' found in the DataFrame.")

In [ ]:
#Keep only 6 variable in fact table to create scatter plot with Gun Per capital since they have week association
imp =['White alone, percent, July 1, 2016,  (V2016)',
    'Persons 65 years and over, percent, April 1, 2010',
    'Owner-occupied housing unit rate, 2011-2015',
    'Asian alone, percent, July 1, 2016,  (V2016)',
    'Foreign born persons, percent, 2011-2015', 
    'Median gross rent, 2011-2015']

In [ ]:
for col in imp:
    plt.figure(figsize=(4,4))
    print(col)
    plt.scatter(fact[col],fact['Gun_Per_Capital_2016'], label =col)
    plt.title(col+" and Gun_Per_Capital Scatter Plot")     
    plt.ylabel("Gun_Per_Capital_2016")
    plt.xlabel(col)
    plt.grid(True)
    plt.show()

# Q2Which states have had the highest growth and the lowest growth in gun registrations? 

In [ ]:
result['gun_growth'] = result['2016_totals']/(result['2016_totals']-result['2010_totals'])


In [ ]:
result['gun_growth'].max() # the result of the biggest growth %


In [ ]:
#List Top 5 rows by gun growth rate descending
result.nlargest(5,'gun_growth')

#Create Bar chart for every states'gun growth 


In [ ]:
  plt.bar(fact.index, fact[col], label=col_str)

In [ ]:
plt.rcdefaults()
fig, ax = plt.subplots()

In [ ]:
sorted = result.sort_values(by=['gun_growth'])


In [ ]:
y_pos = np.arange(len(sorted['state']))
error = np.random.rand(len(sorted['state']))
ax.barh(y_pos, (sorted['gun_growth']*100), xerr=error, align='center',height=2,linewidth=5,color='green', ecolor='black')


In [ ]:
ax.set_yticks(y_pos)
ax.set_xlabel("Gun Registration Growth %")
ax.set_yticklabels(sorted['state'],size=6)

In [ ]:
ax.invert_yaxis()  # labels read top-to-bottom


In [ ]:
ax.tick_params(axis='x', colors='green')
ax.tick_params(axis='y', colors='black')

In [ ]:
plt.title("State Gun Registration Growth from 2010 and 2016")     
plt.grid(True)
plt.legend(bbox_to_anchor=(1.1, 1.05))
plt.show()

Research Question 3  :
# 
# What is the overall trend of gun purchases?

In [ ]:
plt.figure(figsize=(10,5))

gun.groupby('month')['totals'].sum().plot(kind='line',sharex=True, sharey=True, layout = (2, 1))
#set x and y axis lable name
plt.xlabel('Month')
plt.ylabel('Total gun purchases')
plt.legend()
plt.title("Gun Purchase Trend by months Line Chart")
plt.show()

#2. Create line chart using grouped data by year, which can be used to observe the change during the season
plt.figure(figsize=(10,6))
gun.groupby('year')['totals'].sum().plot(kind='line')

plt.ylabel('Total gun purchases')
plt.xlabel('year')
plt.title("Gun Purchase Trend by years Line Chart")
plt.legend()
plt.show()

In [ ]:
!python -m nbconvert --to html Investigate_a_Dataset.ipynb